File: 2408-number-of-people-aware-of-a-secret/number-of-people-aware-of-a-secret.go
const MOD = 1000000007

func peopleAwareOfSecret(n int, delay int, forget int) int {
	// dp[t] = how many people learn the secret on day t
	dp := make([]int, n+1)
	dp[1] = 1 // day 1: one person knows the secret

	for t := 1; t <= n; t++ {
		people := dp[t] // number of people who got the secret on day t

		// they will share from (t+delay) to (t+forget-1)
		for x := t + delay; x < min(t+forget, n+1); x++ {
			dp[x] = (dp[x] + people) % MOD
		}
	}

	// At the end, only those who learned within last `forget-1` days still remember
	res := 0
	for i := n - forget + 1; i <= n; i++ {
		if i >= 1 {
			res = (res + dp[i]) % MOD
		}
	}

	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
    
	return b
}

File: 2408-number-of-people-aware-of-a-secret/number-of-people-aware-of-a-secret.py
MAX = 10**9 + 7

class Solution:
    def peopleAwareOfSecret(self, T: int, delay: int, forget: int) -> int:
        # dp[t] = how many people *learn* the secret on day t
        dp = [0] * (T + 1)
        dp[1] = 1  # on day 1, exactly 1 person knows the secret initially

        # simulate day by day
        for t in range(1, T + 1):
            people = dp[t]  # number of people who got the secret at day t

            # each of these people will start sharing the secret
            # from day (t + delay) until day (t + forget - 1)
            for x in range(t + delay, min(t + forget, T + 1)):
                dp[x] += people  # they pass the secret to new people

        # At the end of day T, some people have already forgotten.
        # Only those who learned the secret within the last `forget - 1` days still remember.
        return sum(dp[-forget:]) % MAX